### PR TITLE
fix build on go 1.18: panic: qtls.ClientHelloInfo doesn't match

### LINKED ISSUE #122: 10-to-11 migration fails with s3ds plugin
I have a custom compiled IPFS image using `v0.8.0` with `v0.7.0` of the `go-ds-s3` plugin bundled in, as per the instructions in `go-ds-s3`'s README. When I try to run the new version after upgrading from `v0.7.0`, I get the following log output:

```
ipfs version 0.8.0
Found IPFS fs-repo at /data/ipfs
Initializing daemon...
go-ipfs version: 0.8.0-ce693d7
Repo version: 11
System version: amd64/linux
Golang version: go1.14.4
Found outdated fs-repo, migrations need to be run.
  => Looking for suitable fs-repo-migrations binary.
  => None found, downloading.
  => Running: /tmp/go-ipfs-migrate889866779/fs-repo-migrations -to 11 -y
Found fs-repo version 10 at /data/ipfs
===> Running migration 10 to 11...
applying 10-to-11 repo migration
  - opening datastore at "/data/ipfs"
ipfs migration:  migration 10 to 11 failed: cannot open datastore: unknown datastore type: s3ds
  => Failed: /tmp/go-ipfs-migrate889866779/fs-repo-migrations -to 11 -y
The migrations of fs-repo failed:
  migration failed: exit status 1
If you think this is a bug, please file an issue and include this whole log output.
  https://github.com/ipfs/fs-repo-migrations
```

The only thing that I was able to find that was similar was ipfs/go-ipfs#6416, but it was closed without being resolved, presumably because the original author disappeared. I tried starting up a completely unmodified node via the `ipfs/go-ipfs` image on Docker Hub, but that had the exact same problem. How am I supposed to run the migration if the migration tool doesn't support the backend that I've got my data stored in?

### LINKED ISSUE #156: Go 1.16 is EOL
Go 1.16 is no longer supported since Go 1.18 was released.
Unfortunately, fs-repo-migrations does not work with Go 1.17. See https://github.com/ipfs/fs-repo-migrations/issues/122#issuecomment-945194121 and https://github.com/NixOS/nixpkgs/commit/2ff6ffe70b8b4051812c83db4753520cec4b56d5 for previous mentions and more detailed explanations of this issue.
Since Go 1.16 is EOL, it will start to be removed in repositories, here for example in NixOS: https://github.com/NixOS/nixpkgs/pull/168480. This will cause fs-repo-migrations to no longer work there.
I would like fs-repo-migrations to be updated to support a current Go version or at least clearly document, which migration is compatible with which Go version.

### LINKED ISSUE #157: panic: qtls.ClientHelloInfo doesn't match
Hello, 

I tried to run the binary `fs-repo-migration` but it panics with the following output:
```
panic: qtls.ClientHelloInfo doesn't match

goroutine 1 [running]:
github.com/ipfs/fs-repo-migrations/ipfs-10-to-11/_vendor/github.com/marten-seemann/qtls-go1-15.init.0()
	/home/peltoche/go/pkg/mod/github.com/ipfs/fs-repo-migrations@v1.7.1/ipfs-10-to-11/_vendor/github.com/marten-seemann/qtls-go1-15/unsafe.go:20 +0x132
```

ipfs version: `0.12.2`
go version: `go1.18.1 linux/amd64`
I have installed `fs-repo-migration` with the following command: `go install github.com/ipfs/fs-repo-migrations@latest`

### PR DESCRIPTION
**Pull request by milahu**
fix #156 
fix #157
fix #122

this (patching `vendor/`) is the simple way
and it works = both forward and reverse migrations finish without error
currently, migrations work from version 7 to 12

<details>

```sh
# warning: revert may cause data loss

path=/home/user/.ipfs

for a in $(seq 11 -1 7); do
  echo a=$a
  ./fs-repo-$a-*/fs-repo-* -path=$path -verbose=true -revert
  sleep 2 # workaround. fixme: migration should be sync
done

for a in $(seq 7 11); do
  echo a=$a
  ./fs-repo-$a-*/fs-repo-* -path=$path -verbose=true
  sleep 2 # workaround. fixme: migration should be sync
done
```

```
a=11
reverting 11-to-12 repo migration
locking repo at "/home/user/.ipfs"
  - verifying version is '12'
  - starting raw multihash to CIDv1 block migration
  - loading repo configurations
  - backup file will be read from /home/user/.ipfs/11-to-12-cids.txt
MFS Root: QmeY5vvgBJED4GsiHe35ubCmeGwiRFkA92gFaH6EEunxzg
2023-01-11 20:31:57.92946784 +0100 CET m=+12.757777763: Generic migration worker syncing after 839 objects migrated
2023-01-11 20:31:57.931535458 +0100 CET m=+12.759845342: Generic migration worker syncing after 839 objects migrated
839 multihashes swapped to CidV1s
reverted version file to version 11
a=10
reverting migration
opening datastore at "/home/user/.ipfs"
reverting pinning to use ipld storage
converted 22 pins from datastore to ipld storage
updated version file
a=9
reverting migration
lowered version number to 9
a=8
reverting migration
Renaming key's filename:  key_gezeim2ln5xvorstmvfvcr3ynzzhqrcljq4gqqlpljifqwtoojjw26tkpbghsqlfgzivordvonbuetkbjbqq
updated version file
a=7
reverting migration
lowered version number to 7
```

```
a=7
applying 7-to-8 repo migration
locking repo at "/home/user/.ipfs"
  - verifying version is '7'
> Upgrading config to new format
updated version file
Migration 7 to 8 succeeded
a=8
applying 8-to-9 repo migration
Renaming key's filename:  12D3KooWFSeKQGxnrxDKL8hAoZPXZnrSmzjxLyAe6QWDusCBMAHa
updated version file
Migration 8 to 9 succeeded
a=9
applying 9-to-10 repo migration
locking repo at "/home/user/.ipfs"
  - verifying version is '9'
> Upgrading config to new format
updated version file
Migration 9 to 10 succeeded
a=10
applying 10-to-11 repo migration
opening datastore at "/home/user/.ipfs"
upgrading pinning to use datastore
converted 22 pins from ipld storage into datastore
updated version file
Migration 10 to 11 succeeded
a=11
applying 11-to-12 repo migration
locking repo at "/home/user/.ipfs"
  - verifying version is '11'
  - loading repo configurations
  - starting CIDv1 to raw multihash block migration
  - backup file will be written to /home/user/.ipfs/11-to-12-cids.txt
  - Adding keys in prefix /blocks to backup file
could not parse /blocks/CEKKA3LXSBZOAHTTLJKUCLOGE2IABGZA5B7Q as a Cid
could not parse /blocks/DMQD5OBY4XKNEVVELXQVQTATPG3SYQJHFF4ZZ4XDZ6TULT345YVWRMI as a Cid
could not parse /blocks/CEKNYI7757GS72XPQA3P3JC6SSNP3BVK4XLQ as a Cid
could not parse /blocks/CEKFAJPXNH2ASKNWXQ27ILO6QRLBH23OWZPQ as a Cid
458 CIDv1 keys added to backup file for /blocks
  - Adding keys in prefix /filestore to backup file
0 CIDv1 keys added to backup file for /filestore
  - backup file will be read from /home/user/.ipfs/11-to-12-cids.txt
2023-01-11 20:32:44.974778209 +0100 CET m=+0.533970302: Migration worker has moved 458 flatfs files and 458 in total
2023-01-11 20:32:44.974941058 +0100 CET m=+0.534133115: Generic migration worker syncing after 0 objects migrated
2023-01-11 20:32:44.974997261 +0100 CET m=+0.534189293: Generic migration worker syncing after 0 objects migrated
458 CidV1s swapped to multihashes
updated version file
```

</details>

7-to-6 fails

```sh
a=6
./fs-repo-$a-*/fs-repo-* -path=$path -verbose=true -revert
# error: expected ed25519 data size to be 96
```

https://github.com/ipfs/fs-repo-migrations/blob/6c9f7fc4268bcff26578a5367038376549bbbcb3/fs-repo-6-to-7/gx/ipfs/QmaPbCnUMBohSGo3KnxEa2bHqyJVVeEEcwtqJAYxerieBo/go-libp2p-crypto/ed25519.go#L122

> this (patching `vendor/`) is the simple way

i also tried the hard way, trying to update the `quic-go` dependency, but failed

<details>

```sh
cd fs-repo-migrations
cd fs-repo-11-to-12

go mod graph | grep github.com/lucas-clemente/quic-go@v0.19.3
# github.com/ipfs/go-ipfs@v0.8.0 github.com/lucas-clemente/quic-go@v0.19.3
# github.com/libp2p/go-libp2p-quic-transport@v0.10.0 github.com/lucas-clemente/quic-go@v0.19.3

# update to quic-go@v0.20.0
go mod graph | grep github.com/lucas-clemente/quic-go
# v0.19.3
go mod edit -replace github.com/lucas-clemente/quic-go=github.com/lucas-clemente/quic-go@v0.20.0
go get github.com/lucas-clemente/quic-go@v0.20.0
go mod vendor
# error
# github.com/libp2p/go-libp2p-quic-transport imports
# github.com/lucas-clemente/quic-go/metrics: no required module provides package github.com/lucas-clemente/quic-go/metrics

# update to go-libp2p-quic-transport@v0.11.0
go mod graph | grep github.com/libp2p/go-libp2p-quic-transport
# v0.10.0
go mod edit -replace github.com/libp2p/go-libp2p-quic-transport=github.com/libp2p/go-libp2p-quic-transport@v0.11.0
go get github.com/libp2p/go-libp2p-quic-transport@v0.11.0
go mod vendor
go build
# error: go-libp2p-quic-transport@v0.11.0 is not compatible with quic-go@v0.20.0
# vendor/github.com/libp2p/go-libp2p-quic-transport/stream.go:12:13: undefined: quic.StreamErrorCode

# rollback
rm -rf vendor/
git reset --hard origin/master

# update to go-libp2p-quic-transport@v0.11.0
go mod graph | grep github.com/libp2p/go-libp2p-quic-transport
# v0.10.0
go mod edit -replace github.com/libp2p/go-libp2p-quic-transport=github.com/libp2p/go-libp2p-quic-transport@v0.11.0
go get github.com/libp2p/go-libp2p-quic-transport@v0.11.0
go mod graph | grep "quic-go@v0.19.3$"
# github.com/ipfs/go-ipfs@v0.8.0 github.com/lucas-clemente/quic-go@v0.19.3
# update to quic-go@v0.21.0
go mod edit -replace github.com/lucas-clemente/quic-go=github.com/lucas-clemente/quic-go@v0.21.0
go get github.com/lucas-clemente/quic-go@v0.21.0
go mod graph | grep "quic-go@v0.19.3$"
# github.com/ipfs/go-ipfs@v0.8.0 github.com/lucas-clemente/quic-go@v0.19.3
# so, either go-ipfs still depends on the old quic-go,
# or this is a bug in "go mod graph"

# update to go-libp2p-quic-transport@v0.12.0
go mod edit -replace github.com/libp2p/go-libp2p-quic-transport=github.com/libp2p/go-libp2p-quic-transport@v0.12.0
go get github.com/libp2p/go-libp2p-quic-transport@v0.12.0
go mod graph | grep quic-go
# github.com/libp2p/go-libp2p-quic-transport@v0.12.0 github.com/lucas-clemente/quic-go@v0.23.0
# update to quic-go@v0.23.0
go mod edit -replace github.com/lucas-clemente/quic-go=github.com/lucas-clemente/quic-go@v0.23.0
go get github.com/lucas-clemente/quic-go@v0.23.0
go mod graph | grep "quic-go@v0.19.3$"
# github.com/ipfs/go-ipfs@v0.8.0 github.com/lucas-clemente/quic-go@v0.19.3
# so, either go-ipfs still depends on the old quic-go,
# or this is a bug in "go mod graph"
go mod vendor
# error: golang.org/x/text/encoding: missing go.sum entry for module providing package golang.org/x/text/encoding (imported by golang.org/x/net/html/charset)
```

</details>

i am aware that migration is rarely needed
but at the same time, its silly to not support such a trivial operation

### CODE REVIEW

File: fs-repo-11-to-12/vendor/github.com/marten-seemann/qtls-go1-15/common.go [hunk 1]
```diff
@@ -430,6 +431,10 @@ type clientHelloInfo struct {
 	// config is embedded by the GetCertificate or GetConfigForClient caller,
 	// for use with SupportsCertificate.
 	config *Config
+
+	// NOTE this field was added in go 1.18
```
Comment (milahu) [inline fs-repo-11-to-12/vendor/github.com/marten-seemann/qtls-go1-15/common.go:435]: or go 1.17?

Comment (milahu) [reply fs-repo-11-to-12/vendor/github.com/marten-seemann/qtls-go1-15/common.go:435]: source
https://github.com/golang/go/blob/6ad27161f8d1b9c5e03fb3415977e1d3c3b11323/src/crypto/tls/common.go#L450

blame
https://github.com/golang/go/commit/860704317e02d699e4e4a24103853c4782d746c1

so this was released in go 1.17

**Comment by Jorropo** [comment]
@milahu we don't maintain older migrations, due to the go crypto lib not exposing the required fields to implement `QUIC`'s TLS we have to use `unsafe` to access private fields of the crypto api.
The crypto API internals may break from version to version so this need manual intervention from the `go-quic` maintainers to update each time.

TL;DR you need to use the go compiler version that was current at the time when we built the migration, while this issue in the go crypto std is not fixed.
I belive https://github.com/lucas-clemente/quic-go/issues/2727 tracks progress on this issue.
